### PR TITLE
Python2.6 compatibility and gcc + castxml fixes

### PR DESCRIPTION
**Pull request by praetorian20**
`parser_test_case.py`
* Monkey-patch `assertIn` and `assertNotIn` methods for Python2.6

`remove_template_defaults.hpp`
* Use `std::tr1` unordered containers for gcc >= 4.4.7 and castxml
* I've tested the following combinations
 * gcc 4.4.7 + gccxml
 * gcc 4.9 + castxml

`autoconfig.py`
* Cosmetic fix for generator name printed to stdout when launching unit tests

### CODE REVIEW

File: unittests/autoconfig.py [hunk 1]
```diff
@@ -56,6 +56,11 @@ class cxx_parsers_cfg(object):
         if 'msvc9' == gccxml.compiler:
             gccxml.define_symbols.append('_HAS_TR1=0')
 
+if cxx_parsers_cfg.gccxml.xml_generator:
```
Comment (praetorian20) [inline unittests/autoconfig.py]: I don't think this change actually affects anything, but I was confused at first to see `Gccxml` being printed despite setting `xml_generator_path` and `xml_generator` to `caslxml` in `xml_generator.cfg`

File: unittests/parser_test_case.py [hunk 2]
```diff
@@ -84,3 +85,10 @@ def _test_calldef_exceptions(self, calldef, exceptions):
             (calldef.name,
              pprint.pformat([delc.name for delc in exception_decls]),
              pprint.pformat([delc.name for delc in exceptions_indeed])))
+
+if sys.version_info < (2, 7, 0):
```
Comment (iMichka) [inline unittests/parser_test_case.py]: If you have unittest2 installed, I think this is not needed; there is a Travis build with python2.6 which is working, so I am wondering why this change is needed.

Comment (praetorian20) [inline unittests/parser_test_case.py]: Ah ok, I don't have unittest2 installed, and was seeing `AttributeError`s in unit tests that used `assertIn()`. Added an additional check to do the above only if unittest2 is not imported.

File: unittests/parser_test_case.py [hunk 3]
```diff
@@ -84,3 +87,10 @@ def _test_calldef_exceptions(self, calldef, exceptions):
             (calldef.name,
              pprint.pformat([delc.name for delc in exception_decls]),
              pprint.pformat([delc.name for delc in exceptions_indeed])))
+
+if sys.version_info < (2, 7, 0) and not found_unittest2:
```
Comment (iMichka) [inline unittests/parser_test_case.py]: I would prefer if we just raise an exception if unittest2 is not installed (that would be 2 lines in the try/except block). Then we can safely rely on the unittest2 library instead of having to monkeypatch stuff.

It's adding a dependency (only for testing) on python 2.6, but I think we can live with it (and I am personally not using python 2.6 anymore; the tests just need to work on Travis).

File: unittests/autoconfig.py (hunk 1 above)
Comment (iMichka) [inline unittests/autoconfig.py]: Not sure how to reproduce what you mean. Could you give me the configuration file you used ?

**Comment by iMichka** [comment]
I have added two comments.

The build passed (https://travis-ci.org/gccxml/pygccxml/builds/113338691), which is positive. I need to look why Travis did not report the result back here.

The changes for the containers are ok (these preprocessor macro definitions are growing out of control :)).

Tell me if you feel confident in squashing all the commits to a single commit. If you want I can just merge it like this once we are done (I know that squashing commits can be a daunting task the first time one needs to do it, so I will not bother you too much with this).

File: unittests/parser_test_case.py (hunk 3 above)
Comment (praetorian20) [inline unittests/parser_test_case.py]: Normally I'd agree with you, but in this case, `unittest2` isn't installed by default on Centos6.7. I'm trying to upgrade a project at work from gccxml to castxml, and I'd like to keep additional dependencies to a minimum. This also the reason I'm using Python2.6 instead of 2.7.

If more and more things show up that require `unittest2`, then absolutely, make that a requirement. But for this particular instance, with a workaround of 6 lines of code, I'd prefer to keep this.

But it's up to you, if you feel strongly that we should throw an exception (removing the `import try-catch`), I can change the code to do that.

File: unittests/autoconfig.py (hunk 1 above)
Comment (praetorian20) [inline unittests/autoconfig.py]: Here's my config file.

```
[gccxml]
xml_generator=castxml
xml_generator_path=/opt/tbs/tbs-CastXML01/root/usr/bin/castxml
compiler=g++
compiler_path=/opt/rh/devtoolset-3/root/usr/bin/g++
```
Without the above change to `autoconfig.py`, I see this output (with DEBUG logging turned on)

```
python test_all.py
Gccxml configured to simulate compiler g++
DEBUG Reading project files: file by file
INFO Parsing source file "core_ns_join_1.hpp" ... 
DEBUG castxml cmd: /opt/tbs/tbs-CastXML01/root/usr/bin/castxml  ...
```
Notice how the second line says `Gccxml` despite castxml being called to do the parsing. With the change, that line becomes
```
Castxml configured to simulate compiler g++
```

**Comment by praetorian20** [comment]
I'll squash the commits into one as soon as I hear back from you about handling `unittest2`

File: unittests/parser_test_case.py (hunk 3 above)
Comment (iMichka) [inline unittests/parser_test_case.py]: I know this kind of situation were you have some legacy libraries (very often at work), so I fully understand.

As this is just for the testing part, it will not affect pygccxml itself. I have no strong opinion so lets be flexible. Please remove unitest2 completely and apply the patch for python < 2.7. That'll do.

File: unittests/autoconfig.py (hunk 1 above)
Comment (iMichka) [inline unittests/autoconfig.py]: Ok I get it.

For the moment, gccxml is still the "default" xml generator. I was rather conservative here, because I wanted people to have time to update their code. I plan to make castxml the default xml generator for release 1.8.0 (and keep gccxml as the second option).

```pygccxml.utils.find_xml_generator()``` makes a wild guess and tries to find a xml generator. If you have gccxml installed it ill pick that up first (by default).
Then, you are overriding the values with the config (which is the recommended way to do it, I just added ```find_xml_generator()``` to help beginners).

So your change is legitimate. Perfect.

**Comment by praetorian20** [comment]
Sorry about closing and reopening, that squash was a bit hairier than I thought it would be :) You have a single commit with all the changes now.

**Comment by iMichka** [comment]
Thank you for the contribution ! This is now merged, and I'll add it to the 1.7.3 release.
